src/recon_buildblock/PLSPrior.cxx

```diff
@@ -0,0 +1,663 @@
+                                continue;
+                       image_gradient_elem[z][y][x]=image[z+1][y][x]- image[z][y][x];
+
+                        }
+                        if(direction==1){
+                            if(y+1>max_y)
+                                continue;
+                       image_gradient_elem[z][y][x]=image[z][y+1][x]- image[z][y][x];
+//                       std::cout<<"grady ="<<image[z][y+1][x]- image[z][y][x]<<std::endl;
+                        }
+                        if(direction==2){
+                            if(x+1>max_x )
+                                continue;
+                       image_gradient_elem[z][y][x]=image[z][y][x+1]- image[z][y][x];
+                        }
+
+                    }
+                 }
+          }
+
+}
+
+template <typename elemT>
+void
+PLSPrior<elemT>::compute_normalisation_anatomical_gradient(DiscretisedDensity<3,elemT> &norm_im_grad,
+                                          const DiscretisedDensity<3,elemT> &image_grad_z,
+                                          const DiscretisedDensity<3,elemT> &image_grad_y,
+                                          const DiscretisedDensity<3,elemT> &image_grad_x){
+
+    const int min_z = image_grad_x.get_min_index();
+    const int max_z = image_grad_x.get_max_index();
+
+
+        for (int z=min_z; z<=max_z; z++)
+          {
+
+            const int min_y = image_grad_x[z].get_min_index();
+            const int max_y = image_grad_x[z].get_max_index();
+
+
+
+              for (int y=min_y;y<= max_y;y++)
+                {
+
+                  const int min_x = image_grad_x[z][y].get_min_index();
+                  const int max_x = image_grad_x[z][y].get_max_index();
+
+
+
+                    for (int x=min_x;x<= max_x;x++)
+                    {
+                     norm_im_grad[z][y][x]   = sqrt (square(image_grad_z[z][y][x]) + square(image_grad_y[z][y][x]) + square(image_grad_x[z][y][x]) + square(this->eta));
+                    }
+                 }
+          }
+}
+
+template <typename elemT>
+void
+PLSPrior<elemT>::
+compute_inner_product_and_penalty(DiscretisedDensity<3,elemT> &inner_product,
+                                  DiscretisedDensity<3,elemT> &penalty,
+                                  DiscretisedDensity<3,elemT> &pet_im_grad_z,
+                                  DiscretisedDensity<3,elemT> &pet_im_grad_y,
+                                  DiscretisedDensity<3,elemT> &pet_im_grad_x,
+                      const DiscretisedDensity<3,elemT> &pet_image){
+
+
+    shared_ptr<DiscretisedDensity<3,elemT> > anatomical_im_grad_z_sptr(this->get_anatomical_grad_sptr (0));
+    shared_ptr<DiscretisedDensity<3,elemT> > anatomical_im_grad_y_sptr(this->get_anatomical_grad_sptr (1));
+    shared_ptr<DiscretisedDensity<3,elemT> > anatomical_im_grad_x_sptr(this->get_anatomical_grad_sptr (2));
+
+    DiscretisedDensity<3,elemT> &norm=*this->anatomical_sptr.get ()->get_empty_copy ();
```

Comment:
this is a memory leak again. You allocate a new object, and never deallocate it properly.j In fact, you're no longer using this, so just get rid of this line.